### PR TITLE
RSDK-4945 PTG bases should only call SetVelocity if the velocities have changed

### PR DESCRIPTION
**Pull request by biotinker**
This PR updates PTG kinematic bases such that a GoToInputs call will only call SetVelocity as needed, when desired velocity changes, rather than calling it constantly on each loop iteration.

### CODE REVIEW

File: components/base/kinematicbase/ptgKinematics.go [hunk 1]
```diff
@@ -157,21 +159,28 @@ func (ptgk *ptgBaseKinematics) GoToInputs(ctx context.Context, inputs []referenc
...
-			linVel,
-			angVel,
-			nil,
-		)
-		if err != nil {
-			return multierr.Combine(err, ptgk.Base.Stop(ctx, nil))
+		// This should call SetVelocity if:
+		// 1) this is the first iteration of the loop, or
+		// 2) either of the linear or angular velocities has changed
+		if i == 0 || !(linVel.ApproxEqual(lastLinVel) && angVel.ApproxEqual(lastAngVel)) {
```
Comment (raybjork) [inline components/base/kinematicbase/ptgKinematics.go:165]: could get rid of the need for the first case if you initialize the last vectors to something other than the zero vector

File: components/base/kinematicbase/ptgKinematics.go [hunk 2]
```diff
@@ -157,21 +159,28 @@ func (ptgk *ptgBaseKinematics) GoToInputs(ctx context.Context, inputs []referenc
...
+			ptgk.logger.Debugf(
+				"setting velocity to linear %v angular %v and running velocity step for %s",
+				linVel,
+				angVel,
+				timestep,
+			)
+
+			err := ptgk.Base.SetVelocity(
```
Comment (raybjork) [inline components/base/kinematicbase/ptgKinematics.go:173]: could bring this into the conditional below

**Review by raybjork — APPROVED** [review]
Thanks for turning this around so fast!  Just a couple optional comments but I'm good to merge this

File: components/base/kinematicbase/ptgKinematics.go (hunk 1 above)
Comment (biotinker) [reply components/base/kinematicbase/ptgKinematics.go:165]: I thought about initializing to `math.Inf(1)` repeated three times but I think this is more readable.

File: components/base/kinematicbase/ptgKinematics.go (hunk 2 above)
Comment (biotinker) [reply components/base/kinematicbase/ptgKinematics.go:173]: Stylistically I personally just don't like the 
```
if x := func(); cond(x) {
   stuff
}
```
paradigm as much as keeping it separate. Easier to tell where the scopes are that way.

File: components/base/kinematicbase/ptgKinematics.go (hunk 1 above)
Comment (biotinker) [reply components/base/kinematicbase/ptgKinematics.go:165]: And there are no other values that we can guarantee will never be called.

Also considered making it a pointer with a nil check, but that's more than keeping track of the loop var that we already have.

Comment (raybjork) [reply components/base/kinematicbase/ptgKinematics.go:165]: I like the infinities personally, keeps you from having to do this logic every iteration. But super minor so whatever you decide to do is good with me

Comment (biotinker) [reply components/base/kinematicbase/ptgKinematics.go:165]: Branch prediction means that this extra conditional will be a no-op.

Comment (nicksanford) [reply components/base/kinematicbase/ptgKinematics.go:165]: I prefer the index test as it makes it clear it is just testing for which iteration of the loop you are on.

File: components/base/kinematicbase/ptgKinematics.go [hunk 3]
```diff
@@ -157,21 +159,28 @@ func (ptgk *ptgBaseKinematics) GoToInputs(ctx context.Context, inputs []referenc
...
+				ctx,
+				linVel,
+				angVel,
+				nil,
+			)
+			if err != nil {
+				return multierr.Combine(err, ptgk.Base.Stop(ctx, nil))
```
Comment (nicksanford) [inline components/base/kinematicbase/ptgKinematics.go:180]: What happens if you do `multierr.Combine(err,nil)`?
Should we test for whether Stop() returned an error before combining it with the setvelocity error?

File: components/base/kinematicbase/ptgKinematics.go (hunk 1 above)
Comment (nicksanford) [inline components/base/kinematicbase/ptgKinematics.go:165]: What are the units we are operating on here? Millimeters? 
If so I take it that that the `1e-16` epsilon that [ApproxEqual](https://github.com/golang/geo/blob/6adc56603217/r3/vector.go#L30) uses is what makes this semantically equivalent to an equality check. 

Is that right?

**Review by nicksanford — APPROVED** [review]
Minor question about existing code. The changes look good though.

File: components/base/kinematicbase/ptgKinematics.go (hunk 3 above)
Comment (biotinker) [reply components/base/kinematicbase/ptgKinematics.go:180]: `multierr.Combine(err,nil)` should result in `err`, or `nil` if `err` is also `nil`.

If we Stop and that returns an error, what would we do aside from just return it?

File: components/base/kinematicbase/ptgKinematics.go (hunk 1 above)
Comment (biotinker) [reply components/base/kinematicbase/ptgKinematics.go:165]: You are correct regarding the equality check.

Units are mm per second for linVel, and degress per second for angVel.

I can add a comment specifying that if you like; the reason there isn't already is I considered the lines above to be self-documenting:
```
		linVel := r3.Vector{0, trajNode.LinVelMMPS, 0}
		angVel := r3.Vector{0, 0, rdkutils.RadToDeg(trajNode.AngVelRPS)}
```

Comment (nicksanford) [reply components/base/kinematicbase/ptgKinematics.go:165]: no need to comment this is good, just confirming my own understanding as I onboard.

File: components/base/kinematicbase/ptgKinematics.go (hunk 3 above)
Comment (nicksanford) [reply components/base/kinematicbase/ptgKinematics.go:180]: This is good in that case. I was just not sure if `multierr.Combine(err,nil)` might result in a multierr with a nil pointer inside of it which might panic when if the Error method is called on it. 

However it sounds like that isn't something we need to worry about.

**Pull request merged**